File: src/benchmark/baselines/FlatBaselineAdapter.js
// FlatBaselineAdapter — E1 control adapter for the academic benchmark.
//
// Purpose: act as the "single LLM, no hierarchy" baseline against the project's
// 4-channel agent (StrategicDirector + EnvironmentDirector + NPCBrain +
// ColonyAgent). Instead of issuing four separate prompts, this adapter fuses
// all four channels into one prompt + one round-trip, then dispatches the
// resulting four sub-objects back through `request(channel, ...)` lookups.
//
// Token budget alignment: each per-channel prompt is constructed by the usual
// PromptPayload builders, then concatenated. The reported per-channel
// `usage.promptTokens` is `Math.round(fusedPromptTokens / 4)` so a 4-channel
// hierarchical baseline and a flat baseline are scored on comparable budgets.
//
// Cache: a single fused response is reused for 30 seconds (configurable via
// `cacheTtlMs`). After expiry the next `request()` triggers another fused
// LLM call. Cache hits do NOT count against `llmCallCount`.
//
// Fallback: if the LLM call fails, the schema parse fails, or any sub-channel
// is missing, the adapter returns a `fallback: true` DecisionResponse with
// `data: null` for the affected channel. The remaining channels still see
// whatever fields parsed successfully.
//
// Test seam: pass `llmClient` directly via `opts.llmClient` to inject a
// stubbed object exposing `requestStrategic(prompt, enabled, fallbackData)`.

import { AgentAdapter, CHANNELS, SCHEMA_VERSION } from "../../simulation/ai/llm/AgentAdapter.js";
import { LLMClient } from "../../simulation/ai/llm/LLMClient.js";
import {
  buildEnvironmentPromptUserContent,
  buildPolicyPromptUserContent,
} from "../../simulation/ai/llm/PromptPayload.js";
import { guardEnvironmentDirective, guardGroupPolicies } from "../../simulation/ai/llm/Guardrails.js";
import { validateEnvironmentDirective, validateGroupPolicy } from "../../simulation/ai/llm/ResponseSchema.js";

const DEFAULT_CACHE_TTL_MS = 30_000;
const DEFAULT_TIMEOUT_MS = 30_000;

const CHANNEL_KEY = Object.freeze({
  "environment-director": "env",
  "npc-policy": "policy",
  "strategic-plan": "strategic",
  "colony-agent": "colony",
});

function nowMs() {
  return (typeof performance !== "undefined" && typeof performance.now === "function")
    ? performance.now()
    : Date.now();
}

function emptyDirectiveForChannel(channel, error) {
  return {
    data: null,
    fallback: true,
    usage: { promptTokens: 0, completionTokens: 0, cachedTokens: 0 },
    latencyMs: 0,
    model: "flat-baseline-fallback",
    error: String(error || "flat baseline fallback"),
    debug: { channel, reason: "fallback" },
  };
}

// Crude word-count token estimate; matches what the rest of the benchmark
// uses when usage.promptTokens isn't provided by the proxy.
function estimateTokens(text) {
  if (!text) return 0;
  const s = String(text);
  if (!s.trim()) return 0;
  // ~4 chars per token, OpenAI heuristic.
  return Math.max(1, Math.round(s.length / 4));
}

function buildFusedPrompt(payloadByChannel) {
  const envPrompt = payloadByChannel["environment-director"]
    ? buildEnvironmentPromptUserContent(payloadByChannel["environment-director"])
    : "";
  const policyPrompt = payloadByChannel["npc-policy"]
    ? buildPolicyPromptUserContent(payloadByChannel["npc-policy"])
    : "";
  const strategicSummary = payloadByChannel["strategic-plan"] ?? null;
  const colonySummary = payloadByChannel["colony-agent"] ?? null;

  const header = [
    "Make all 4 decisions at once: weather (env), policy (groups), strategic plan, colony build.",
    "Output JSON object {env: {...}, policy: {...}, strategic: {...}, colony: {...}}.",
    "Each sub-object must validate against its channel schema; do not emit prose outside JSON.",
  ].join("\n");

  const body = {
    schemaVersion: SCHEMA_VERSION,
    fused: true,
    instructions: header,
    channels: {
      env: envPrompt || null,
      policy: policyPrompt || null,
      strategic: strategicSummary,
      colony: colonySummary,
    },
  };

  return JSON.stringify(body, null, 2);
}

function parseFusedResponse(raw) {
  if (raw && typeof raw === "object" && !Array.isArray(raw)) {
    // Already an object — accept either {env,policy,strategic,colony} or
    // a wrapper {data: {...}}.
    if (raw.env || raw.policy || raw.strategic || raw.colony) return raw;
    if (raw.data && typeof raw.data === "object") return raw.data;
    return raw;
  }
  if (typeof raw !== "string") return null;
  try {
    return JSON.parse(raw);
  } catch {
    return null;
  }
}

export class FlatBaselineAdapter extends AgentAdapter {
  /**
   * @param {object} [opts]
   * @param {string} [opts.baseUrl]
   * @param {string} [opts.model]
   * @param {number} [opts.timeoutMs]
   * @param {number} [opts.cacheTtlMs]
   * @param {object} [opts.llmClient] — injected stub for tests; must expose
   *   `requestStrategic(promptStr, enabled, fallbackData)` returning
   *   `{fallback, data, latencyMs, error, model, debug}`.
   */
  constructor(opts = {}) {
    super();
    this.baseUrl = opts.baseUrl ?? "";
    this.model = opts.model ?? "";
    this.timeoutMs = Number(opts.timeoutMs) > 0 ? Number(opts.timeoutMs) : DEFAULT_TIMEOUT_MS;
    this.cacheTtlMs = Number(opts.cacheTtlMs) > 0 ? Number(opts.cacheTtlMs) : DEFAULT_CACHE_TTL_MS;
    this.llmClient = opts.llmClient ?? new LLMClient({ baseUrl: this.baseUrl });

    // Cache state: shared across all 4 channels.
    this._cache = null; // { expiresAtMs, parsed, model, fusedPromptTokens, completionTokens, latencyMs }
    this._pendingPayloads = new Map(); // channel -> latest payload (waiting to be fused)
    this.llmCallCount = 0;
    this.cacheHitCount = 0;
  }

  _isCacheFresh() {
    return this._cache !== null && this._cache.expiresAtMs > nowMs();
  }

  _invalidateCache() {
    this._cache = null;
  }

  async _fuseAndCall() {
    const payloadByChannel = {};
    for (const ch of CHANNELS) {
      const p = this._pendingPayloads.get(ch);
      if (p !== undefined) payloadByChannel[ch] = p;
    }
    const fusedPrompt = buildFusedPrompt(payloadByChannel);
    const fusedPromptTokens = estimateTokens(fusedPrompt);

    let result;
    try {
      result = await this.llmClient.requestStrategic(fusedPrompt, true, null);
    } catch (err) {
      result = {
        fallback: true,
        data: null,
        latencyMs: 0,
        error: String(err?.message ?? err ?? "llm error"),
        model: this.model || "flat-baseline-error",
        debug: null,
      };
    }
    this.llmCallCount += 1;

    const parsed = parseFusedResponse(result?.data);
    const completionTokens = estimateTokens(typeof result?.data === "string" ? result.data : JSON.stringify(result?.data ?? ""));

    this._cache = {
      expiresAtMs: nowMs() + this.cacheTtlMs,
      parsed: parsed && typeof parsed === "object" ? parsed : null,
      model: String(result?.model ?? this.model ?? "flat-baseline"),
      fusedPromptTokens,
      completionTokens,
      latencyMs: Number(result?.latencyMs) || 0,
      fallback: Boolean(result?.fallback) || parsed === null,
      error: String(result?.error ?? ""),
      debug: result?.debug ?? null,
    };
    return this._cache;
  }

  /**
   * @param {string} channel
   * @param {object} payload
   * @param {object} [options]
   * @returns {Promise<object>} DecisionResponse
   */
  async request(channel, payload, options = {}) {
    if (!CHANNELS.includes(channel)) {
      return emptyDirectiveForChannel(channel, `unknown channel: ${channel}`);
    }
    // Update pending payload for this channel — only used when next fused
    // call fires (cache miss).
    this._pendingPayloads.set(channel, payload);

    let cache = this._cache;
    if (!this._isCacheFresh()) {
      cache = await this._fuseAndCall();
    } else {
      this.cacheHitCount += 1;
    }
    if (!cache) {
      return emptyDirectiveForChannel(channel, "fused-call returned no cache");
    }

    const key = CHANNEL_KEY[channel];
    const subPayload = cache.parsed ? cache.parsed[key] : null;

    // Per-channel schema gate — env + policy validate; strategic/colony pass
    // through (their schemas live in ColonyPlanner / StrategicDirector and
    // depend on state context not available here).
    let validatedData = subPayload ?? null;
    let fallback = cache.fallback || subPayload == null;
    let error = cache.error;

    if (!fallback) {
      if (channel === "environment-director") {
        const v = validateEnvironmentDirective(subPayload);
        if (!v.ok) { fallback = true; error = `schema: ${v.error}`; validatedData = null; }
        else validatedData = guardEnvironmentDirective(v.value);
      } else if (channel === "npc-policy") {
        // policy validator expects {policies:[...]} envelope.
        const candidate = subPayload.policies ? subPayload : { policies: subPayload };
        const v = validateGroupPolicy(candidate);
        if (!v.ok) { fallback = true; error = `schema: ${v.error}`; validatedData = null; }
        else validatedData = guardGroupPolicies(v.value);
      }
      // strategic-plan + colony-agent: caller validates downstream.
    }

    if (fallback && validatedData == null) {
      return {
        ...emptyDirectiveForChannel(channel, error || "fused parse failed"),
        latencyMs: cache.latencyMs,
        model: cache.model,
        usage: {
          promptTokens: Math.round(cache.fusedPromptTokens / 4),
          completionTokens: Math.round(cache.completionTokens / 4),
          cachedTokens: 0,
        },
      };
    }

    return {
      data: validatedData,
      fallback,
      usage: {
        // Aligned with 4-channel separate-call path: divide fused prompt/
        // completion budget by 4.
        promptTokens: Math.round(cache.fusedPromptTokens / 4),
        completionTokens: Math.round(cache.completionTokens / 4),
        cachedTokens: 0,
      },
      latencyMs: cache.latencyMs,
      model: cache.model,
      error,
      debug: {
        channel,
        fused: true,
        cacheAgeMs: Math.max(0, this.cacheTtlMs - (cache.expiresAtMs - nowMs())),
        llmCallCount: this.llmCallCount,
        cacheHitCount: this.cacheHitCount,
      },
    };
  }
}

export default FlatBaselineAdapter;

File: project_utopia/benchmark/baselines/flat_baseline_adapter.py
"""FlatBaselineAdapter — E1 control adapter (single LLM, no hierarchy).

Python port of ``src/benchmark/baselines/FlatBaselineAdapter.js``.

Behaviour:

- Single fused prompt for all 4 channels per cache window
  (default 30 s TTL).
- Each ``request(channel, payload)`` returns the per-channel slice of the
  fused response with prompt/completion token budgets divided by 4 so
  hierarchical and flat baselines are compared on equal budgets.
- Per-channel schema gates run on ``environment-director`` and
  ``npc-policy``; ``strategic-plan`` and ``colony-agent`` pass through.
- Failure modes (transport error, JSON parse failure, schema reject, missing
  sub-channel) return a ``fallback=True`` :class:`DecisionResponse`.
"""

from __future__ import annotations

import json
import time
from typing import Any

from pydantic import ValidationError

from ...simulation.ai.llm.agent_adapter import (
    AgentAdapter,
    CHANNELS,
    DecisionResponse,
    SCHEMA_VERSION,
    UsageStats,
)
from ...simulation.ai.llm.guardrails import (
    guard_environment_directive,
    guard_group_policies,
)
from ...simulation.ai.llm.prompt_payload import (
    build_environment_prompt_user_content,
    build_policy_prompt_user_content,
)
from ...simulation.ai.llm.response_schema import EnvironmentDirective, NpcPolicyEnvelope

DEFAULT_CACHE_TTL_MS: float = 30_000.0
DEFAULT_TIMEOUT_MS: float = 30_000.0

CHANNEL_KEY: dict[str, str] = {
    "environment-director": "env",
    "npc-policy": "policy",
    "strategic-plan": "strategic",
    "colony-agent": "colony",
}


def _now_ms() -> float:
    return time.monotonic() * 1000.0


def _estimate_tokens(text: Any) -> int:
    """Crude ~4-chars-per-token heuristic (matches the JS port)."""
    if text is None:
        return 0
    s = str(text)
    if not s.strip():
        return 0
    return max(1, round(len(s) / 4))


def _build_fused_prompt(payload_by_channel: dict[str, Any]) -> str:
    env_prompt = (
        build_environment_prompt_user_content(payload_by_channel["environment-director"])
        if "environment-director" in payload_by_channel and payload_by_channel["environment-director"]
        else ""
    )
    policy_prompt = (
        build_policy_prompt_user_content(payload_by_channel["npc-policy"])
        if "npc-policy" in payload_by_channel and payload_by_channel["npc-policy"]
        else ""
    )
    strategic_summary = payload_by_channel.get("strategic-plan")
    colony_summary = payload_by_channel.get("colony-agent")

    header = "\n".join((
        "Make all 4 decisions at once: weather (env), policy (groups), strategic plan, colony build.",
        "Output JSON object {env: {...}, policy: {...}, strategic: {...}, colony: {...}}.",
        "Each sub-object must validate against its channel schema; do not emit prose outside JSON.",
    ))
    body = {
        "schemaVersion": SCHEMA_VERSION,
        "fused": True,
        "instructions": header,
        "channels": {
            "env": env_prompt or None,
            "policy": policy_prompt or None,
            "strategic": strategic_summary,
            "colony": colony_summary,
        },
    }
    return json.dumps(body, indent=2, sort_keys=True)


def _parse_fused_response(raw: Any) -> dict[str, Any] | None:
    if isinstance(raw, dict):
        # Accept either {env, policy, strategic, colony} directly, or {"data": {...}}.
        if any(k in raw for k in ("env", "policy", "strategic", "colony")):
            return raw
        if isinstance(raw.get("data"), dict):
            return raw["data"]
        return raw
    if not isinstance(raw, str):
        return None
    try:
        return json.loads(raw)
    except (json.JSONDecodeError, ValueError):
        return None


def _empty_directive_for_channel(channel: str, error: str | Exception | None) -> DecisionResponse:
    return DecisionResponse(
        data=None,
        fallback=True,
        usage=UsageStats(prompt_tokens=0, completion_tokens=0, cached_tokens=0),
        latency_ms=0.0,
        model="flat-baseline-fallback",
        error=str(error or "flat baseline fallback"),
        debug={"channel": channel, "reason": "fallback"},
    )


class _StubLLMResult(dict):
    """Container used internally for the stub LLM client return value."""


class FlatBaselineAdapter(AgentAdapter):
    """Fused 4-channel adapter (E1 control)."""

    def __init__(
        self,
        *,
        base_url: str = "",
        model: str = "",
        timeout_ms: float = DEFAULT_TIMEOUT_MS,
        cache_ttl_ms: float = DEFAULT_CACHE_TTL_MS,
        llm_client: Any = None,
    ) -> None:
        super().__init__()
        self.base_url = base_url
        self.model = model
        self.timeout_ms = float(timeout_ms) if float(timeout_ms) > 0 else DEFAULT_TIMEOUT_MS
        self.cache_ttl_ms = float(cache_ttl_ms) if float(cache_ttl_ms) > 0 else DEFAULT_CACHE_TTL_MS
        self.llm_client = llm_client  # Must expose async `request_strategic(prompt, enabled, fallback_data)`.

        self._cache: dict[str, Any] | None = None
        self._pending_payloads: dict[str, Any] = {}
        self.llm_call_count: int = 0
        self.cache_hit_count: int = 0

    def _is_cache_fresh(self) -> bool:
        return self._cache is not None and float(self._cache["expires_at_ms"]) > _now_ms()

    def _invalidate_cache(self) -> None:
        self._cache = None

    async def _fuse_and_call(self) -> dict[str, Any]:
        payload_by_channel: dict[str, Any] = {}
        for ch in CHANNELS:
            if ch in self._pending_payloads:
                payload_by_channel[ch] = self._pending_payloads[ch]
        fused_prompt = _build_fused_prompt(payload_by_channel)
        fused_prompt_tokens = _estimate_tokens(fused_prompt)

        if self.llm_client is None or not hasattr(self.llm_client, "request_strategic"):
            result = {
                "fallback": True,
                "data": None,
                "latency_ms": 0.0,
                "error": "no llm client",
                "model": self.model or "flat-baseline-error",
                "debug": None,
            }
        else:
            try:
                result = await self.llm_client.request_strategic(fused_prompt, True, None)
            except Exception as err:  # noqa: BLE001 — translate to fallback envelope
                result = {
                    "fallback": True,
                    "data": None,
                    "latency_ms": 0.0,
                    "error": str(err) or "llm error",
                    "model": self.model or "flat-baseline-error",
                    "debug": None,
                }
        self.llm_call_count += 1

        parsed = _parse_fused_response(result.get("data") if isinstance(result, dict) else None)
        raw_data = result.get("data") if isinstance(result, dict) else None
        completion_text = raw_data if isinstance(raw_data, str) else json.dumps(raw_data or "")
        completion_tokens = _estimate_tokens(completion_text)

        self._cache = {
            "expires_at_ms": _now_ms() + self.cache_ttl_ms,
            "parsed": parsed if isinstance(parsed, dict) else None,
            "model": str(result.get("model") if isinstance(result, dict) else (self.model or "flat-baseline")),
            "fused_prompt_tokens": int(fused_prompt_tokens),
            "completion_tokens": int(completion_tokens),
            "latency_ms": float(result.get("latency_ms", 0) if isinstance(result, dict) else 0) or 0.0,
            "fallback": bool(result.get("fallback") if isinstance(result, dict) else True) or parsed is None,
            "error": str(result.get("error", "") if isinstance(result, dict) else ""),
            "debug": result.get("debug") if isinstance(result, dict) else None,
        }
        return self._cache

    async def request(
        self,
        channel: str,
        payload: Any,
        options: dict[str, Any] | None = None,
    ) -> DecisionResponse:
        if channel not in CHANNELS:
            return _empty_directive_for_channel(channel, f"unknown channel: {channel}")
        self._pending_payloads[channel] = payload

        cache = self._cache
        if not self._is_cache_fresh():
            cache = await self._fuse_and_call()
        else:
            self.cache_hit_count += 1
        if not cache:
            return _empty_directive_for_channel(channel, "fused-call returned no cache")

        key = CHANNEL_KEY[channel]
        sub_payload = cache["parsed"].get(key) if isinstance(cache["parsed"], dict) else None

        validated_data: Any = sub_payload
        fallback = bool(cache["fallback"]) or sub_payload is None
        error = cache["error"]

        if not fallback:
            if channel == "environment-director":
                try:
                    validated = EnvironmentDirective.model_validate(sub_payload)
                    validated_data = guard_environment_directive(validated).model_dump(by_alias=False)
                except (ValidationError, ValueError) as ve:
                    fallback = True
                    error = f"schema: {ve}"
                    validated_data = None
            elif channel == "npc-policy":
                candidate = sub_payload if isinstance(sub_payload, dict) and "policies" in sub_payload else {"policies": sub_payload}
                try:
                    validated = NpcPolicyEnvelope.model_validate(candidate)
                    validated_data = guard_group_policies(validated).model_dump(by_alias=False)
                except (ValidationError, ValueError) as ve:
                    fallback = True
                    error = f"schema: {ve}"
                    validated_data = None

        per_channel_prompt = max(0, round(cache["fused_prompt_tokens"] / 4))
        per_channel_completion = max(0, round(cache["completion_tokens"] / 4))

        if fallback and validated_data is None:
            empty = _empty_directive_for_channel(channel, error or "fused parse failed")
            return DecisionResponse(
                data=None,
                fallback=True,
                usage=UsageStats(
                    prompt_tokens=per_channel_prompt,
                    completion_tokens=per_channel_completion,
                    cached_tokens=0,
                ),
                latency_ms=float(cache["latency_ms"]),
                model=str(cache["model"]),
                error=empty.error,
                debug={"channel": channel, "reason": "fallback"},
            )

        return DecisionResponse(
            data=validated_data,
            fallback=fallback,
            usage=UsageStats(
                prompt_tokens=per_channel_prompt,
                completion_tokens=per_channel_completion,
                cached_tokens=0,
            ),
            latency_ms=float(cache["latency_ms"]),
            model=str(cache["model"]),
            error=error,
            debug={
                "channel": channel,
                "fused": True,
                "cache_age_ms": max(0.0, self.cache_ttl_ms - (float(cache["expires_at_ms"]) - _now_ms())),
                "llm_call_count": self.llm_call_count,
                "cache_hit_count": self.cache_hit_count,
            },
        )


__all__ = ["FlatBaselineAdapter"]
